app/src/main/java/org/cru/godtools/ui/tooldetails/ToolDetailsFragment.kt

```diff
@@ -131,16 +149,21 @@ class ToolDetailsFragment() :
         val code = tool?.code ?: return
         val locale = translation?.languageCode ?: return
 
-        activity?.run {
-            openToolActivity(code, tool.type, locale, showTips = true)
-            if (settings.getFeatureDiscoveredCount("$FEATURE_TUTORIAL_TIPS$code") < 3) {
-                settings.setFeatureDiscovered("$FEATURE_TUTORIAL_TIPS$code")
-                startTutorialActivity(
-                    PageSet.TIPS,
-                    Bundle().apply { putString(FORMAT_ARG_TOOL_NAME, translation.getName(tool).toString()) }
-                )
-            }
-        }
+        dataModel.tipsLanguage = locale
+        dataModel.tipsTool = code
+        dataModel.tipsType = tool.type
+        launchTrainingTips(code, tool.type, locale, false)
+    }
+
+    private fun launchTrainingTips(
+        code: String,
+        type: Tool.Type,
+        locale: Locale,
+        force: Boolean
+    ) {
+        if (force || settings.isFeatureDiscovered("$FEATURE_TUTORIAL_TIPS$code"))
+            requireActivity().openToolActivity(code, type, locale, showTips = true)
+        else startActivityForResult(context?.buildTutorialActivityIntent(PageSet.TIPS), REQUEST_TUTORIAL_TIPS)
```

Comment:
change `context?` to `requireContext()`

Reply:
I assume ```requireContext``` just makes sure the context is available before running the function?

Reply:
yeah, `requireContext()` returns a non-null context or throws an error. without that change, when `context` is null, you will have a null intent, which would throw a different error when trying to use `startActivityForResult`.

Reply:
similar concept as `requireActivity()` vs `activity`